src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -16,14 +16,235 @@
 #include "native_client/src/shared/platform/nacl_sync_checked.h"
 
 #include "native_client/src/trusted/service_runtime/arch/sel_ldr_arch.h"
+#include "native_client/src/trusted/service_runtime/dyn_array.h"
+#include "native_client/src/trusted/service_runtime/nacl_app.h"
 #include "native_client/src/trusted/service_runtime/nacl_desc_effector_ldr.h"
 #include "native_client/src/trusted/service_runtime/nacl_globals.h"
-#include "native_client/src/trusted/service_runtime/nacl_tls.h"
-#include "native_client/src/trusted/service_runtime/nacl_switch_to_app.h"
 #include "native_client/src/trusted/service_runtime/nacl_stack_safety.h"
+#include "native_client/src/trusted/service_runtime/nacl_switch_to_app.h"
 #include "native_client/src/trusted/service_runtime/nacl_syscall_common.h"
+#include "native_client/src/trusted/service_runtime/nacl_tls.h"
 #include "native_client/src/trusted/service_runtime/osx/mach_thread_map.h"
 
+#include "native_client/src/include/win/mman.h"
+#include "native_client/src/trusted/service_runtime/load_file.h"
+#include "native_client/src/trusted/service_runtime/sel_memory.h"
+
+#include "native_client/src/trusted/desc/nacl_desc_io.h"
+#include "native_client/src/shared/platform/lind_platform.h"
+#include "native_client/src/trusted/service_runtime/include/bits/mman.h"
+#include "native_client/src/trusted/service_runtime/include/sys/fcntl.h"
+
+
+/*
+ * always points at original program context
+ *
+ * -jp
+ */
+struct NaClThreadContext *master_ctx;
+
+
+/*
+ * dynamically allocate and initilize a copy
+ * of the parents NaClApp structure which is
+ * used in NaClSysFork()
+ */
+struct NaClApp *NaClChildNapCtor(struct NaClAppThread *natp) {
+  struct NaClApp *nap_child = NaClAlignedMalloc(sizeof *nap_child, __alignof(struct NaClApp));
+  struct NaClApp *nap_master = ((struct NaClAppThread *)master_ctx)->nap;
+  struct NaClApp *nap_parent = natp->nap;
+  NaClErrorCode *mod_status = NULL;
+
+  CHECK(nap_master);
```

Comment:
What `CHECK` is performed here? (I can look it up, but perhaps the macro name can guide my understanding.)

Reply:
This seems to do pretty much the same, including spelling out the checked condition:

```c
#include <assert.h>
int main(int argc, char** argv) {
  assert ( argc == 0 );
  return 0;
}
```
Compile and run (OS X 10.11.6, clang-800.0.42.1):
```
Assertion failed: (argc == 0), function main, file foo.c, line 3.
Abort trap: 6
```

Reply:
This is just an assert() basically, but the rest of the codebase uses CHECK() so I was keeping things consistent.